packages/flutter_tools/lib/src/run.dart

```diff
@@ -114,67 +116,114 @@ class RunAndStayResident extends ResidentRunner {
       }
     }
 
-    _package = getApplicationPackageForPlatform(device.platform, applicationBinary: applicationBinary);
+    _result = null;
+    Stopwatch startTime = new Stopwatch()..start();
 
-    if (_package == null) {
-      String message = 'No application found for ${device.platform}.';
-      String hint = getMissingPackageHintForPlatform(device.platform);
-      if (hint != null)
-        message += '\n$hint';
-      printError(message);
-      return 1;
+    String toolsExtension = Cache.instance.getToolsExtension('run');
+    if (toolsExtension != null) {
```

Comment:
This whole block of code should be factored into a function and the else block should be factored into a function as well.